airbyte-integrations/connectors/source-blob/source_blob/filereader.py

```diff
@@ -0,0 +1,144 @@
+            "number": ("float64", "float16", "float32", "decimal128", "decimal256", "halffloat", "float", "double"),
+            "string": ("large_string", "string"),
+            "object": ("large_string", ),
+            "array": ("large_string", ),
+            "null": ("large_string", )
+        }
+        if not reverse:
+            for json_type, pyarrow_types in map.items():
+                if str_typ.lower() == json_type:
+                    return getattr(pa, pyarrow_types[0]).__call__()  # better way might be necessary when we decide to handle more type complexity
+            logger.warn(f"JSON type '{str_typ}' is unknown, falling back to default conversion to large_string")
+            return pa.large_string()
+        else:
+            for json_type, pyarrow_types in map.items():
+                if any([str_typ.startswith(pa_type) for pa_type in pyarrow_types]):
+                    return json_type
+            logger.warn(f"PyArrow type '{str_typ}' is unknown, falling back to default conversion to string")
+            return "string"  # default type if unspecified in map
+
+    @staticmethod
+    def json_schema_to_pyarrow_schema(schema: Mapping[str, Any], reverse: bool = False) -> Mapping[str, Any]:
+        """ TODO docstring """
+        new_schema = {}
+
+        for column, json_type in schema.items():
+            new_schema[column] = FileReader.json_type_to_pyarrow_type(json_type, reverse=reverse)
+
+        return new_schema
+
+
+class FileReaderCsv(FileReader):
+    """  TODO Docstring """
+    pass
+
+    @property
+    def is_binary(self):
+        return True
+
+    def _read_options(self):
+        return pa.csv.ReadOptions(
+            block_size=10000,
+            encoding=self._format.get("encoding", 'utf8')
+        )
+
+    def _parse_options(self):
+        return pa.csv.ParseOptions(
+            delimiter=self._format.get("delimiter", ','),
+            quote_char=self._format.get("quote_char", '"'),
+            double_quote=self._format.get("double_quote", True),
+            escape_char=self._format.get("escape_char", False),
+            newlines_in_values=self._format.get("newlines_in_values", False)
+        )
+
+    def _convert_options(self, json_schema=None):
+        check_utf8 = True if self._format.get("encoding", 'utf8').lower().replace("-", "") == 'utf8' else False
+        convert_schema = self.json_schema_to_pyarrow_schema(json_schema) if json_schema is not None else None
+        return pa.csv.ConvertOptions(
+            check_utf8=check_utf8,
+            column_types=convert_schema,
+            **json.loads(self._format.get("additional_reader_options", '{}'))
+        )
+
+    def get_inferred_schema(self, file) -> dict:
+        streaming_reader = pa_csv.open_csv(file, self._read_options(), self._parse_options(), self._convert_options())
+        schema_dict = {field.name: field.type for field in streaming_reader.schema}
```

Comment:
how does pa determine the type of fields in a CSV? Does it have to read the full CSV? (that would be bad)

Reply:
Since we're using [open_csv()](https://arrow.apache.org/docs/python/generated/pyarrow.csv.open_csv.html) rather than [read_csv()](https://arrow.apache.org/docs/python/generated/pyarrow.csv.read_csv.html), this creates a StreamingReader here and infers the schema by only reading the first block ([described here](https://github.com/apache/arrow/blob/master/docs/source/python/csv.rst#incremental-reading) in caveat 2)